Navigation Block: Add Color Options for Submenus

georgeh opened the pull request:
Closes #29963

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Adds options to color the submenus separately from the main navigation block

### TODO

- [x] Add contrast checking
- [x] Get colors working on dynamic blocks like Page List

## How has this been tested?
Manually tested

## Screenshots <!-- if applicable -->
https://user-images.githubusercontent.com/128240/115888059-7ee86400-a420-11eb-9b5a-76ffb065f137.mov

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Navigation Block: Add Color Options for Submenus

Closes #29963

georgeh commented:
@jasmussen - One thought I had about this is that the color menu becomes really big. We could hide the overlay color options until there are submenus, but that would also make the color options depend on the state of the block tree. I'm worried that would be confusing. Do you have any thoughts on whether it's better to make the overlay colors dynamic or always there?

jasmussen commented:
This is awesome! Thanks so much for working on this. This is what I see:

![this](https://user-images.githubusercontent.com/1204802/116043170-99fce300-a66f-11eb-9bca-0f542f7e827e.gif)

This is working well, except for two things:

1. When you set custom text or background colors for the overlay menu, they don't work. I didn't quite figure out why, but in case I have a moment later, I'll take a look.
2. We should change the text labels. For the menu overall, the labels should be `Text` and `Background`, and for overlays it should be `Overlay text` and `Overlay background`.

There's a separate bug I appear to have introduced, where hovering submenus when the parent doesn't have a background isn't as easy as it used to. I'll fix this separately. Similarly, the new margin/padding rules that I introduced which zero out the paddings when a background color is present that needs a little massaging:

<img width="904" alt="Screenshot 2021-04-26 at 09 12 08" src="https://user-images.githubusercontent.com/1204802/116043956-a03f8f00-a670-11eb-9cd8-bb4a0fa251ec.png">

I will take a look at that too when I have a moment.

To your question:

> One thought I had about this is that the color menu becomes really big. We could hide the overlay color options until there are submenus, but that would also make the color options depend on the state of the block tree. I'm worried that would be confusing. Do you have any thoughts on whether it's better to make the overlay colors dynamic or always there?

Yes, we should not fix that here. Or to put it differently, it's okay that this menu is big because the issue has to be fixed separately. Specifically as part of https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866 and https://github.com/WordPress/gutenberg/issues/27331

Commit:
Fix custom overlay colors

Commit:
Fix menu labels

jasmussen commented:
After your latest rebase and tweaks, I see this:

![currently](https://user-images.githubusercontent.com/1204802/116205364-f8909280-a73d-11eb-8982-a0e0bd603f4c.gif)

https://github.com/WordPress/gutenberg/pull/31195 contains a fix for the submenu offset shown at the end of the GIF, so that's fixed separately.

While I do agree about the sidebar colors, I remain convinced that it is something we should fix separately as part of  #27473 or #27331. @gziolo would you agree?

Given that, I think this one is ready for review!

gziolo commented:
> While I do agree about the sidebar colors, I remain convinced that it is something we should fix separately as part of #27473 or #27331. @gziolo would you agree?

That's a good point. When there is more granular control over all settings we call Global Styles, it will be easier to reason what to render based on the current conditions, or maybe to gray out parts that don't apply at the moment.

I only glanced at the code, but It seems to look good. I like the direction 👍🏻

jasmussen commented:
Just to be clear, what I was hoping to revisit separately was to, instead of showing this:

<img width="292" alt="Screenshot 2021-04-27 at 10 25 46" src="https://user-images.githubusercontent.com/1204802/116210279-ff6dd400-a742-11eb-9173-a1994477c349.png">

Then show a different interface for selecting colors, something more like this:

<img width="633" alt="Screenshot 2021-04-27 at 10 26 41" src="https://user-images.githubusercontent.com/1204802/116210364-14e2fe00-a743-11eb-9c2d-a2753d3ede06.png">

georgeh commented:
I'm going to do what I can to get contrast checks working, then I'll get this ready to merge.

georgeh commented:
Contrast checking is _kinda_ working, but there's no good way to indicate which colors are actually conflicting without rebuilding the whole color panel.
![image](https://user-images.githubusercontent.com/128240/116264971-b6208300-a748-11eb-8f65-19ba9e8df07d.png)

That would be a pretty big change and a lot of duplicate code. Is it better to remove contrast checking or ship it like this while we wait for #27331?

Commit:
Add Contrast Checking

jasmussen commented:
Great questions. I see these as indications that we need to get started on the unified color picker very soon. 

From testing it, to me it looks like contrast checking between nav foreground and background is working, but not overlay foreground and background:

<img width="365" alt="Screenshot 2021-04-28 at 09 54 17" src="https://user-images.githubusercontent.com/1204802/116367684-e4b16300-a807-11eb-92c7-632cdb27deda.png">

yes it would be nice if the contrast checker message was shown more in context of the two colors it's comparing, such as below colors 1 and 2. An alternative would be to show it above, but then we'd want to rephrase to be less verbose as it's currently a very big message.

If there are no regressions compared to trunk, I think this feature is totally fine to ship as is, especially if we create followup tickets for issues and crosslink them well with the tickets noted. What do you think?

georgeh commented:
@jasmussen I am able to get both contrast warnings to appear. The only scenario I can think of where it wouldn't trigger is if you don't have any submenus in the navigation, because it needs a DOM node to get the computed style. If you _do_ have submenus in the nav you're testing, I would appreciate hearing about that. 

We also cannot move the contrast checker's position, due to how `PanelColorSettings` is implemented, so we can't put one above and one below or between the colors. `PanelColorSettings` renders the `ContrastChecker` as a child. It might be possible to rework the message to add a subject, e.g. "**The overlay color combination** may be hard for people to read" instead of "This color combination may be hard for people to read"

This is ready for review, I'll merge when I get the ✅  and create issues for:
- Improve contrast messaging
- Show contrast checker on overlay without needing a submenu in the nav
- Revisit color controls once the core controls improve

Am I missing any?

jasmussen commented:
Great call. I had submenus that were generated by the Page List:

![colors](https://user-images.githubusercontent.com/1204802/116520066-856c5500-a8d2-11eb-816a-63ec653c065d.gif)

In that case, it doesn't work.

But if I convert them:

![convert to blocks](https://user-images.githubusercontent.com/1204802/116520083-8a310900-a8d2-11eb-8cf8-11297c691a81.gif)

jasmussen commented:
I'm not seeing the overlay colors take effect on the frontend, and I'm seeing an issue where it isn't working with the Page List block.

georgeh commented:
I'm running into a blocker with getting the Page List to render the colors. Right now I'm using the context to render classes/inline CSS and it works fine on the rendered page. However, the way the editor renders dynamic blocks means that we don't get the context. The result is that in the Page List block, the submenus inherit the colors.

This problem has been brought up in #28684, which was inspired by the Page List block.

Should we clean this up and ship it as-is, with the submenu colors working in the context of Page Link blocks in both editor and front-end, and Page List blocks on the front-end?

I considered CSS vars, but they won't work for theme colors' class names. My other thought is exploring some way to convert the context into attributes so that when the editor requests the block, it will get the colors from the context.

Commit:
Add Submenu Colors to Page List Block

Commit:
Clean up some redundant styles

jasmussen commented:
Looking into this, not sure I can help fix, but I'd like to try, and in that process I rebased and pushed. I'll return.

jasmussen commented:
Alright I see the problem.

We are applying things like "has-yellow-background" classes directly onto the submenu items — which is appropriate enough — but that breaks with the Page List block because that's a server side rendered block mostly.

Two ideas for moving forward, I'll let you all decide what's best:

- Rewrite Page List as a client-block [per advice here](https://github.com/WordPress/gutenberg/issues/28684#issuecomment-772459370). Kind of a reboot.
- Change the approach here to, instead of assigning classes to submenus, instead assigning a "has-yellow-overlay-background-color" on the navigation block itself. That way it can be inherited.

What do you think? CC: @gwwar

georgeh commented:
I'm leaning towards client block as a solution, thanks for that pointer. I'd rather take on the burden of making a responsive block instead of pushing the burden on to themers.

jasmussen commented:
Cc @mkaz as we just chatted about the page list also.

georgeh commented:
@jasmussen @gwwar If I can get the client block working I'll open a PR for just that change, and then this can follow-up on that. I'm making enough progress that I think this is a good way to go forward.

gwwar commented:
> Two ideas for moving forward, I'll let you all decide what's best:
> Rewrite Page List as a client-block per advice here. Kind of a reboot.
> Change the approach here to, instead of assigning classes to submenus, instead assigning a "has-yellow-overlay-background-color" on the navigation block itself. That way it can be inherited.

So one thing we'd need to do before attempting to rewrite page list as a client block, is doing a performance profiling pass on navigation links. Right now the editor visibility slows down with only > 50 or so which would very quickly show up on page list if we swapped implementations.

gwwar commented:
To help move this forward, since we can now transform small lists of PageList -> NavigationLinks, I think it'd be fine to reduce scope here to target NavigationLinks only.

jasmussen commented:
Another option is to also apply the color classes to the navigation block itself. That way they can be inherited in CSS only.

jasmussen commented:
@georgeh thinking about this some more, perhaps it really is best that the overlay menu color property is a property of _the navigation block itself_ rather than _each individual menu item_. Because it might be best to use this same color control for the 🍔 menu background, and for that to work, it would need to be inherited.

Commit:
PHP lint fixes

jasmussen commented:
I realized that submenu items can't `inherit` a color of a parent container unless that container has the same color defined and _used_. This is what we do for menu items currently — if the parent container has a blue background, we can use `background-color: inherit;` on submenu items, and they get the same background color. But (obviously in hindsight) the whole point of this overlay color is have a transparent background on the main menu item, and have the colors only on submenu items.

gwwar commented on packages/block-library/src/page-list/index.php:
```diff
@@ -88,28 +127,40 @@ function block_core_page_list_build_css_font_sizes( $context ) {
+			}
+		}
+
+		$markup .= '<li class="' . $css_class . '"' . $style_attribute . '>';
```
Is this working for folks? I'm not seeing the style attribute being added, so the submenus are inheriting from the navigation block.

<img width="1552" alt="Screen Shot 2021-05-18 at 1 59 03 PM" src="https://user-images.githubusercontent.com/1270189/118723299-1489d000-b7e2-11eb-8929-9562c1b49847.png">

gwwar commented:
This is looking pretty close @georgeh! Here's what I see in various editors:

🤔 Widgets Editor: I'm not sure if these are theme bugs, but I don't see overlay colors/text consistently apply. There may also be differences on the published view

https://user-images.githubusercontent.com/1270189/118725036-400dba00-b7e4-11eb-9673-44bff35f5136.mp4

❌  Navigation Editor doesn't persist color settings. Did we want to show this option in this editor, or see if it's possible to add serialization support for saving to menu items?

https://user-images.githubusercontent.com/1270189/118725639-0f7a5000-b7e5-11eb-875c-693ebbb676f7.mp4

SiteEditor + PostEditor: 

* Things work well for NavigationLinks in editor. Published view doesn't appear to be picking up overlay colors.
* Colors for Overlay text/Overlay background don't appear to be set for PageList in editor/published

<img width="1407" alt="Screen Shot 2021-05-18 at 2 33 26 PM" src="https://user-images.githubusercontent.com/1270189/118726437-29686280-b7e6-11eb-95b8-32eae6f98e8a.png">

Commit:
Fix front-end rendering of colors

georgeh commented:
@gwwr @jasmussen can you take another look at this? Getting the editor to display the colors for the Page List block is out of scope for this particular PR. 

Also of note is the way I'm [copying colors from the Navigation block's context into the Navigation Link's attributes so they'll render](https://github.com/WordPress/gutenberg/pull/31149/files#diff-161a9737e181e56e07148b7eee4795f129aa821f73993a276cf1fd9c17941b70R305-R350). I'm still using context because Page List has access to it on the server when rendering, but I'd be curious to hear other ways to get the context in Navigation Link's `save()`.

jasmussen commented:
Took it for a spin, looking like it's close:

![alsohey](https://user-images.githubusercontent.com/1204802/119775381-cfd9f500-bec3-11eb-91d0-c11f864b5ff3.gif)

The fact that the text color doesn't affect the links appears also in trunk, I will look at that separately.

So it looks like the remaining issue is that this doesn't appear to be rendering the overlay color on the frontend. It doesn't look like the right classes are being applied:

<img width="1538" alt="Screenshot 2021-05-27 at 08 18 53" src="https://user-images.githubusercontent.com/1204802/119775655-33fcb900-bec4-11eb-8f28-b6427d842001.png">

CC: @gwwar as I think your other ping didn't take ;)

gwwar commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -295,6 +302,52 @@ export default function NavigationLinkEdit( {
+			backgroundColorAttr = context.backgroundColor,
+			customTextColorAttr = context.customTextColor || style?.color?.text,
+			customBackgroundAttr =
+				context.customBackgroundColor || style?.color?.background;
```
This is pretty optional, but since we're using these properties to assign in `setAttributes` it might make more sense to make this an object.

eg
```jsx
const updatedColors = {
    textColor: context.textColor,
    backgroundColor: context.backgroundColor,
    customTextColor: context.customTextColor ?? style?.color?.text,
    customBackgroundColor: context.customBackgroundColor ?? style?.color?.background
};
//...
setAttributes( updatedColors );
```

georgeh replied:
Thanks, much cleaner!

gwwar commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -44,6 +47,38 @@ const LAYOUT = {
+}
+
+function detectColors( clientId, setColor, setBackground ) {
+	const colorsDetectionElement = getBlockDOMNode( clientId, document );
```
Reading the code, this stands out a bit, do we use this approach anywhere else? Did we run into any trouble with using child refs?

georgeh replied:
This is pretty much copy/pasted from `packages/block-library/src/button/color-edit.js`. I proposed some ideas around making that more extensible but the consensus was that it was better to do a customized implementation within the Navigation block.

I did find [this conversation](https://github.com/WordPress/gutenberg/commit/0c6e36977f32b5ea184cdb7e7e94bac3efb73b34#r43840820) about the original code.

I'll try to see if I can get the colors with a ref.

georgeh replied:
Seems to be working with a ref!

gwwar commented:
Maybe @nosolosw could help take a look too? I think there might be similar color problems here and with global styles.

gwwar commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -102,6 +158,31 @@ function Navigation( {
+	const [ detectedOverlayColor, setDetectedOverlayColor ] = useState();
+
+	useEffect( () => {
+		if ( ! enableContrastChecking ) {
```
This feels more like a block "supports" feature. https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/. Maybe @ajlende or @youknowriad might have thoughts.

Specifically something like:
```
supports: {
    color: {
        __experimentalContrastChecking: true,
    }
}
```

Is there any reason why we didn't update block.json to add supports: color. (color.background and color.text default to true on the object).

For dynamic blocks there are helpers in block-supports folder. https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php

georgeh replied:
> This feels more like a block "supports" feature.

Do you mean `enableContrastChecking`? That's a variable that right now just checks if the platform is web, since mobile doesn't support contrast checking yet.

> Is there any reason why we didn't update block.json to add supports: color. (color.background and color.text default to true on the object).

This is more code that was copied from `packages/block-library/src/button/color-edit.js`, going off [this conversation](https://github.com/WordPress/gutenberg/issues/29963#issuecomment-823132744) to implement a customized version for this block instead of adding filters to the Color Panel.

gwwar replied:
> Do you mean enableContrastChecking?

More that if folks find this feature useful, that others can enable it for other blocks without copy pasting.

> This is more code that was copied from packages/block-library/src/button/color-edit.js, going off this conversation to implement a customized version for this block instead of adding filters to the Color Panel.

Thanks for the extra context! Reading the conversation it sounds like we'd like to explore a bit and observe use cases. One thing that may be helpful for future us, is creating a follow up issue to articulate the concerns there.

gwwar reviewed (commented):
@georgeh let me know when you've rebased, and I can try doing a manual testing pass.

Commit:
Lint

Commit:
Typo

gwwar commented:
Reserving this comment for manual testing

#### Site Editor / Post Editor

Overlay Text/Background not being shown in published view:

| Editor | Published |
|-----|-----|
| <img width="1410" alt="Screen Shot 2021-05-27 at 12 45 04 PM" src="https://user-images.githubusercontent.com/1270189/119888186-ea26c980-bee9-11eb-816e-867931e2b206.png"> | <img width="1410" alt="Screen Shot 2021-05-27 at 12 45 18 PM" src="https://user-images.githubusercontent.com/1270189/119888227-f579f500-bee9-11eb-8ccb-e095ee48a75b.png"> |

It's possible to get multiple contrast warnings. Not blocking but it'd be nice to at least give a hint of which combo was an issue.
<img width="265" alt="Screen Shot 2021-05-27 at 12 44 48 PM" src="https://user-images.githubusercontent.com/1270189/119888256-ff035d00-bee9-11eb-9f06-9898932d4e06.png">

#### Widgets Editor 

Testing with Twenty Fifteen, colors aren't applied on the published page.
| Editor | Published |
|-----|-----|
| <img width="1083" alt="Screen Shot 2021-05-27 at 12 59 43 PM" src="https://user-images.githubusercontent.com/1270189/119889547-8e5d4000-beeb-11eb-9ff1-92782fc52f68.png"> | <img width="1083" alt="Screen Shot 2021-05-27 at 12 59 53 PM" src="https://user-images.githubusercontent.com/1270189/119889569-94532100-beeb-11eb-91fa-2c0b76920c5e.png"> |


#### Navigation Editor

Colors are not persisted. (Try setting some then refreshing the editor).

https://user-images.githubusercontent.com/1270189/119890408-a71a2580-beec-11eb-9692-e8e18925d8a7.mp4

Since this isn't serialized, this also means that colors are not shown in customizer/published view.

| Customizer | Published |
|-----|-----|
| <img width="1083" alt="Screen Shot 2021-05-27 at 1 04 07 PM" src="https://user-images.githubusercontent.com/1270189/119890548-d2047980-beec-11eb-9812-58a8fc0ca6ff.png"> | <img width="980" alt="Screen Shot 2021-05-27 at 1 04 55 PM" src="https://user-images.githubusercontent.com/1270189/119890561-d6c92d80-beec-11eb-9775-8ce31a03fbd6.png"> |

gwwar commented:
@georgeh thanks for working on this one! To help give this a bit of momentum I'd recommend seeing if any of the following steps might make sense.

- [ ] Get a fix in for overlay/background text on the site/post editor. This might be its own PR if there's some issues in trunk.
- [ ] Double check with folks that the copy/paste approach feels okay here. Eg did we need to create any follow up tasks to keep an eye on things?
- [ ] See if we need to support color options in the widget/navigation editors. We can potentially reduce scope further in this PR by only showing the controls on the post/site editor to start with.

Commit:
Fix front-end colors for Navigation Link block

Commit:
Refactor code for easier understanding

jasmussen commented:
Note to self: in a separate followup PR I need to remove the border when a submenu item has a background color. Right now that border is only removed if the _parent_ block has a background color.

Commit:
Page List: Add overlay colors to the editor view

Commit:
Lint

Commit:
Lint

Commit:
Fix tests

jasmussen commented:
I just took this for a spin, and as far as I can tell, this is now working 🚀 

![overlay](https://user-images.githubusercontent.com/1204802/121477296-f588f300-c9c7-11eb-82fa-f234994c97dd.gif)

Frontend:

<img width="850" alt="Screenshot 2021-06-10 at 08 42 46" src="https://user-images.githubusercontent.com/1204802/121477299-f6ba2000-c9c7-11eb-81aa-645ed81c79c7.png">

Commit:
Fix E2E tests

Commit:
Navigation Link: Refactor color attributes

Commit:
Unit test fixtures

mkaz committed:
Update snapshot for nav overlay test

Commit:
Update snapshots

Commit:
Update snapshots

Commit:
Navigation Link: Fix Submenu Colors

Commit:
Lint

georgeh commented:
@gziolo would you mind taking a look the code at this? I duplicated a lot of code from `packages/block-editor/src/hooks/color.js` in `packages/block-library/src/navigation/edit.js`, including contrast checking. I would love to have the color settings in the hook, similar to the paragraph block's link colors using `supports: { colors: { link: true } }`, maybe that is a future refactor?

Are there any changes you'd like to see before this gets merged?

jasmussen commented:
Thanks for continuing to work on this. It's feels close for the basic use case and that works excellently well:

![overlay](https://user-images.githubusercontent.com/1204802/122529223-85a8e700-d01d-11eb-8178-fdb01b675b24.gif)

There are two issues. First one, the parent navigation block colors bleed into the submenu, even when you've set submenu colors:

![overlay submenu color](https://user-images.githubusercontent.com/1204802/122529295-9d806b00-d01d-11eb-8916-80a866e760a7.gif)

You can see here how the unordered list indicating the submenu has the "blinding red color" CSS class:

<img width="1811" alt="Screenshot 2021-06-18 at 10 10 48" src="https://user-images.githubusercontent.com/1204802/122529330-a96c2d00-d01d-11eb-8cf8-c2f3bcf80479.png">

Ideally that menu itself was blue same as the menu items. It's more visible with the mobile menu, where the parent block color is the only color that applies:

<img width="495" alt="Screenshot 2021-06-18 at 10 12 52" src="https://user-images.githubusercontent.com/1204802/122529546-e3d5ca00-d01d-11eb-9c3c-1b9a2086aeb1.png">

I'm afraid that if I remove the background color from the parent block entirely, but keep the overlay color, I get white items (i.e. the white text color I chose is applied, but the blue background is removed). That's actually my fault:

<img width="1811" alt="Screenshot 2021-06-18 at 10 15 41" src="https://user-images.githubusercontent.com/1204802/122529867-36af8180-d01e-11eb-9c57-4bdd48b07446.png">

See the "background: transparent !important" value there? I'd like to push some polish to this branch to help that along, we can probably split this in two, with a small fix letting us land this branch. But ideally the overlay color should take precedence in the mobile menu. Sorry for not catching this sooner, and if you have time next week, let's pair on it. I think we can probably address this one reasonably quickly.

The other issue is the foreground color goes missing when you apply both an overlay background color, and a navigation block background color, but no foreground color. In this case, my default text color for menu items is black, so it should remain black, not turn red. This seems like a smallish issue? Observe:

<img width="2404" alt="Screenshot 2021-06-18 at 10 22 36" src="https://user-images.githubusercontent.com/1204802/122530781-251aa980-d01f-11eb-9d85-55f99adb2bf3.png">

Here we see that even though I applied the red color as a _background_ for the navigation block, the class applied was `has-blinding-red-color` (when it should've been `has-blinding-red-background-color`). That's combined with the red background color also being set as an inline style on the nav element.

Commit:
Fix menu color bleed

jasmussen committed:
Push text inheritance rule.

jasmussen commented:
Nice work, this is nearly there:

![yo](https://user-images.githubusercontent.com/1204802/122883360-696aba00-d33d-11eb-8a62-7a46d472b78f.gif)

"has-blinding-red-color" is still applied to the main menu in addition to "has-blinding-red-background-color" when only the latter should be applied. But otherwise, this is closing in. 

There's two tasks on mobile still:

![mobile](https://user-images.githubusercontent.com/1204802/122884276-44c31200-d33e-11eb-8833-2ccadad7c51e.gif)

Here, the main background is forcefully applied to all menu items. I pushed a fix so that the foreground _text color_ is applied as well. In a separate PR, we can look at using the overlay color for the menu instead of the background color: I think this one is big enough as is.

Because I pushed that color inheritance fix, the one remaining task we need to land is to remove the extra foreground color that's applied. Observe in this screenshot how the burger menu icon is red, on a red background, and in the markup you can see how both the `has-pure-white-color` and `has-blinding-red-color` classes, and it should have only the former. 

The step to reproduce is to simply insert a navigation block with a few items, and apply a background color. The background color is applied as both background and foreground color.

Once that one's done, I think we can land this as a big step on the path.

gziolo commented:
> @gziolo would you mind taking a look the code at this? I duplicated a lot of code from `packages/block-editor/src/hooks/color.js` in `packages/block-library/src/navigation/edit.js`, including contrast checking. I would love to have the color settings in the hook, similar to the paragraph block's link colors using `supports: { colors: { link: true } }`, maybe that is a future refactor?

@jorgefilipecosta and @nosolosw, would you mind doing the check here. You have a much better picture of how colors integration with blocks works at the moment and how the proposed approach fits in the long-term plan.

Commit:
Navigation Link: Fix overlay color bug

jasmussen commented:
Saw your latest commit come in, and didn't see a comment from you so I'm not sure if you were ready for another look. But I took one regardless 😅 

... And I'm still seeing the text color applied to the nav block when only the background should be:

![overlay](https://user-images.githubusercontent.com/1204802/123051083-ad26f780-d401-11eb-9a33-6942ad74e496.gif)

Let me know when you need a look!

Commit:
Fix disappearing text when no text-color specified

georgeh commented:
@jasmussen I think those two issues should be fixed with the last two commits. Please review [048f9e1](https://github.com/WordPress/gutenberg/pull/31149/commits/048f9e19ef611eacd236b6ebd504d314f33bc50f) especially since it's CSS to say "if there's no specified text-color, ignore the rule that says `color: currentColor` and just use `var(--global--color-primary)`"

It seems to match what happens on the front-end but I would appreciate a second set of eyes.

jasmussen commented:
> Please review 048f9e1 especially since it's CSS to say "if there's no specified text-color, ignore the rule that says color: currentColor and just use var(--global--color-primary)"

I think that variable may be coming from one of the Twenty themes, and not something we can count on, not entirely sure. Can you test with Empty Theme from https://github.com/WordPress/theme-experiments so we're sure it's not theme-dependant?

I can test in a bit, but from looking at the CSS I don't think we should employ `:not` here as it adds a great deal of specificity. The problem I was having, with red text on a red background appered to be due to the _text_ color being explicitly set, even when only the background was set. I had both foreground and background classes added, even when I only added the background class. I believe this is different in trunk.

jasmussen commented:
Oh, and thanks for your patience on this one!

Commit:
Revert "Fix disappearing text when no text-color specified"

This reverts commit 048f9e19ef611eacd236b6ebd504d314f33bc50f.

Commit:
Fix text color when cleared

georgeh commented:
@jasmussen good catch, I reverted the CSS changes and found the logic bug in the edit component.

jasmussen commented:
Gosh dang amazing work. This is 99% there:

![currently](https://user-images.githubusercontent.com/1204802/123213966-f4c38700-d4c6-11eb-83b2-6ce050a61e6e.gif)

Two things:

1. The menu now doesn't work on the frontend. This issue is also present in `trunk`, so not the fault of this branch. But CC: @vcanales and @gziolo as I think they both touched how frontend JS is enqueued recently.
2. We need to follow up with separately making the overlay bg and text colors take precedence over the nav block bg/foreground when inside the mobile burger menu. Definitely fine for now.

Very nice work, thanks for staying on this.

jasmussen reviewed (approved):
This is a good user experience and works well for me. The issues noted are existing or followups.

I'd appreciate a pair of eyes on this by @gwwar or @vcanales if they have time, but otherwise looks good!

oandregal commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -70,7 +114,15 @@ function Navigation( {
 			'is-vertical': attributes.orientation === 'vertical',
 			'is-responsive': attributes.isResponsive,
+			'has-text-color': !! textColor.color || !! textColor?.class,
+			[ `has-${ textColor?.slug }-color` ]: !! textColor?.slug,
```
Could you use the `getColorClassName` utility here to create the class from the slug? Otherwise, this code would need to use lodash's `kebabCase` to process the color slug. Same for the background color class.

We're tied to lodash's kebabCase implementation because that utility has been in use since the very beginning to create classes from slugs. [This](https://github.com/WordPress/gutenberg/issues/32347) describes a number of edge cases we need to cover. We had to replicate the [same behavior](https://github.com/WordPress/gutenberg/pull/32766) in the server.

A use case where this would break would be:

- theme provides one color with slug `5red`
- in any block that uses the hooks system or the `getColorClassName` utility the class will be created as `has-5-red-color`
- in navigation will be `has-5red-color`

georgeh replied:
Thanks, I didn't know about that.

oandregal commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -387,26 +471,43 @@ export default function NavigationLinkEdit( {
-			'has-background': !! backgroundColor || !! style?.color?.background,
-			[ `has-${ backgroundColor }-background-color` ]: !! backgroundColor,
+			'has-text-color': !! textColor || !! customTextColor,
+			[ `has-${ textColor }-color` ]: !! textColor && ! customTextColor,
```
Same as https://github.com/WordPress/gutenberg/pull/31149/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#r657796626

oandregal commented on packages/block-library/src/page-list/index.php:
```diff
@@ -9,51 +9,91 @@
 	if ( $has_named_text_color ) {
 		// Add the color class.
-		$colors['css_classes'][] = sprintf( 'has-%s-color', $context['textColor'] );
+		$colors['css_classes'][] = sprintf( 'has-%s-color', $attributes['textColor'] );
```
Same as https://github.com/WordPress/gutenberg/pull/31149/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#r657796626 There's the `gutenberg_experimental_to_kebab_case` function that you can use here.

Same for background and the overlay colors below.

oandregal commented:
>  would you mind doing the check here. You have a much better picture of how colors integration with blocks works at the moment and how the proposed approach fits in the long-term plan.

In terms of direction, this is in line with what we discussed https://github.com/WordPress/gutenberg/issues/29963 (not using the supports but create custom controls), so it's fine.

Commit:
Navigation: use `getColorClassName` to make classnames

Commit:
Lint

georgeh commented:
@gwwar I thought I was able to reproduce the colors not showing up in the Navigation Editor, but now I can't. Here's my process:

1. `npm run wp-env destroy && npm run wp-env start --update`
2. Import [themeunittestdata.wordpress.xml](https://raw.githubusercontent.com/WPTT/theme-test-data/master/themeunittestdata.wordpress.xml)
3. Activate tt1-blocks
4. Go to Site Editor
5. In the Header's navigation block, Add Existing Menu -> Testing Menu (This adds Navigation Link blocks, I've also tested this with "All Pages" to add a Page List block)
6. Set the overlay colors, verify in the editor
7. Save
8. View site

I'm expecting the submenus to show up with the colors on the rendered site, and they are. For a while they were showing up in the editor but not the rendered site. As I tried to create a reproducible test case, the bug stopped happening, so I suspect I wasn't saving.

Am I looking at the right thing? If so, can you give it a test and see if it's working for you? If not can you point me to what needs fixing? Thanks.

gwwar commented:
Sure, so what I'm seeing is that when we visit `/wp-admin/admin.php?page=gutenberg-navigation` the color controls are available on the navigation block, but do not persist (try saving then refreshing the editor).

What I expect here is to not see the color controls OR that these styles are saved and shown correctly in the navigation editor and published page.

https://user-images.githubusercontent.com/1270189/124035716-c4da1f00-d9b1-11eb-9ef0-786be9556d7e.mp4

jorgefilipecosta commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -296,6 +361,26 @@ export default function NavigationLinkEdit( {
 
+	// Store the colors from context as attributes for rendering
+	useEffect( () => {
+		setAttributes( {
```
I think this may be causing undo issues, e.g: if I apply a background overlay color pressing undo does not undo the color as expected. I think the set attributes call is probably creating undo levels.
__unstableMarkNextChangeAsNotPersistent may be helpful in this case.
What makes us need to set the context as attributes? Couldn't we also rely on accessing the context on the server for rendering?

georgeh replied:
Thanks, I think this was vestigial and I was able to simplify it in a414a98989 without having to use `__unstableMarkNextChangeAsNotPersistent`

jorgefilipecosta reviewed (commented):
Hi @georgeh thank you for the work on this PR, things work well in my tests 👍 

I'm wondering how to integrate submenu colors with global styles/theme.json. It would be nice if on the theme.json we could configure the submenu colors for navigation.
It seems like submenu/overlays could be an element of the navigation block (similar to link color that is an element of paragraphs headings).
In that case, global styles would only output something like:
```
.wp-navigation-block .overlay-element-selector {
	color: overlay text color;
	background: overlay background color;
}
```
Would an output like that work?

Commit:
Initial work on Navigation Editor colors

Limitations:
- doesn't seem to support all theme colors, just very basic colors
- does not support custom colors (I don't think there is a place to
  store custom CSS, see wp_setup_nav_menu_item()
- may have trouble with translating classes like-has 5-red-color to 5red
  because we don't know if the hyphen was inserted by us or the theme

tellthemachines commented:
I gave this a quick test in the Navigation editor and I'm seeing the color and background-color classnames getting added to the block elements when setting colors, but they're never removed when new ones are added, so likely whichever color comes last in the stylesheet is the one getting applied:

<img width="427" alt="Screen Shot 2021-07-05 at 3 01 36 pm" src="https://user-images.githubusercontent.com/8096000/124420052-16175380-dda2-11eb-9324-52815e5bf5d5.png">

This is the same in the editor and the front end, but the actual colors can only be seen on the front end because we're not enqueuing theme stylesheets in the Navigation editor. (At this point in time, support for block functionality is limited in this editor because we need back compat with legacy menus, so we're not aiming for a 1:1 visual correspondence between editor and front end. With that in mind, perhaps we should leave adding color support to this editor for a later stage.)

draganescu commented on packages/block-library/src/navigation-link/editor.scss:
```diff
@@ -64,6 +64,12 @@
 		margin-left: $grid-unit-20;
 	}
+
+	// Override the list reset for colored, nested menus
```
Could we add to the comment "why" this override is required?

georgeh replied:
It was supposed to fix a color blowout, but that was fixed elsewhere and I was able to remove the code in f04646fb3f

draganescu commented:
I don't think this should be available in the navigation editor without opt in. By deafault, the navigation editor should respect what the theme does with the menu, because the initial state of this is a drop in replacement of the current WordPress navigation screen.

For themes that opt in to use block menus, that is a different story. So, for now this feature of the navigation block is yet [another feature to be removed](https://github.com/WordPress/gutenberg/issues/30007).

jorgefilipecosta commented:
Hi @draganescu,
> I don't think this should be available in the navigation editor without opt in. By deafault, the navigation editor should respect what the theme does with the menu

Well if by default the block respects the styles the theme is setting we can still allow the user to choose overlay colors, of course, if the user chooses an overlay color we stop respecting the theme but that is expected. It seems the functionality could be available without opt-in but if the user does not choose a color we should respect the theme styles.

jasmussen commented:
I wanted to clarify a small aspect of the block color behavior, as I sense it's pertinent to discussion on whether to allow or not on the navigation screen.

By default, dropdown menus have explicit colors set in the block: white background and black text. This is because the navigation block itself has a transparent background (to work over photo backgrounds), so there's no color for dropdown menus to `inherit`. So we _have_ to choose a default color for dropdown menus, or the menu items inside would be illegible when over the wrong background.

<img width="728" alt="Screenshot 2021-07-06 at 08 31 58" src="https://user-images.githubusercontent.com/1204802/124553130-a69e7400-de34-11eb-8246-37d335fb11ac.png">

When a navigation block _does_ have a background color, we can safely inherit them:

<img width="718" alt="Screenshot 2021-07-06 at 08 33 02" src="https://user-images.githubusercontent.com/1204802/124553235-ca61ba00-de34-11eb-9558-1a5a35cb4a86.png">

This PR adds the ability to set _just the dropdown background color_, enabling this:

<img width="683" alt="Screenshot 2021-07-06 at 08 35 12" src="https://user-images.githubusercontent.com/1204802/124553497-19a7ea80-de35-11eb-8983-3b29482cd719.png">

Without that option, I imagine a theme could still change _just the dropdown background color_ through a complex CSS selector. But I also think that selector would end up with such specificity that it would break the ability for users to override it if the block was used anywhere else but the screen. On the flipside, if the dropdown color was surfaced as just a property to change in theme.json, even if the explicit color control wasn't surfaced in the inspector, I imagine it would be much simpler for themes to customize it.

Commit:
Navigation Editor: Fix for multiple color classes getting added

Commit:
Lint

Commit:
Fix undo for colors and simplify attributes

georgeh commented:
>  It seems like submenu/overlays could be an element of the navigation block (similar to link color that is an element of paragraphs headings).
> In that case, global styles would only output something like:
> 
> ```
> .wp-navigation-block .overlay-element-selector {
> 	color: overlay text color;
> 	background: overlay background color;
> }
> ```
> 
> Would an output like that work?

@jorgefilipecosta I don't see how that will work in the case where there are two navigation blocks on the page with different overlay menu colors. If there was a way to dynamically add an id to the selector when the page saves, that would probably work (and save a bit of bandwidth), but I'm not aware of a good way to do that.

Commit:
Fix unit tests

Commit:
Update E2E specs

Commit:
Remove old CSS

Commit:
E2E test snapshots

jasmussen commented:
Checks are green! @draganescu @jorgefilipecosta what do we need to do to alleviate the concerns of landing this one? Thank you.

georgeh commented:
@jasmussen before landing this, I want to disable custom colors on the Navigation Editor screen, since they don't work.

The Navigation Editor does not store block attributes. Instead, it maps the block attrs to the [fields on nav menu item](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L786). I was able to hack together something to get color slugs from the CSS classes, but there's nowhere to store custom colors.

Also the named colors are not displaying in the editor when selected, but that should be solvable.

jasmussen commented:
👌

gwwar commented:
@georgeh Feel free to ping me when you're ready for another round of testing 👍 I think it makes sense to defer adding the color feature to the Navigation Editor for now. 

There's also a general conversation around how much we should couple the navigation editor to the navigation block that may affect what features we'd like to support.

jorgefilipecosta commented:
> @jorgefilipecosta I don't see how that will work in the case where there are two navigation blocks on the page with different overlay menu colors. 


Hi @georgeh,

Global styles provide styles that are common to all the blocks so yes applying colors to navigation using theme.json would make all menus with the same color.
If there is a need for different colors per block then the theme would create block-specific styles at the block level. Like currently we can change the color of paragraphs globally and at the block level.


> If there was a way to dynamically add an id to the selector when the page saves, that would probably work (and save a bit of bandwidth), but I'm not aware of a good way to do that.

That's exactly what our elements mechanism does, for example, it is used for link color and generates a unique id per link color e.g: `p.wp-element-23232 a { color: link color }`. And we already have hooks that deal with injecting the id's etc. I think that this mechanism would fit very well the use-case if we had a submenu element (and as a bonus things would work well with theme.json right away). That mechanism would also easily allow things like setting a different font size for the submenus etc.

But this is not a blocker I guess for now we can keep this custom control approach and then see if elements are in fact a good fit (if there is a need to change later it will involve creating a deprecation to migrate our current attributes to elements).

Commit:
Remove color picker from Navigation Editor

Commit:
Fix empty classnames

Commit:
Typo

oandregal commented:
> I'm wondering how to integrate submenu colors with global styles/theme.json. It would be nice if on the theme.json we could configure the submenu colors for navigation. It seems like submenu/overlays could be an element of the navigation block (similar to link color that is an element of paragraphs headings).

Hey, wanted to comment on this and reinforce what I said at https://github.com/WordPress/gutenberg/pull/31149#issuecomment-867515659 I consider it fine for this PR to use custom controls, although that means it can't be styled via global styles at the moment (either the sidebar or `theme.json`).

This is an issue that has come up for a couple of complex blocks and we don't want to block those PRs until this use case is supported. I've gathered what we know so far at https://github.com/WordPress/gutenberg/issues/33255 to start thinking about how to solve that issue. While promising, using the elements API requires us to understand first how/whether a UI can be built for them in the post editor, which we don't have designs for yet (this is a concern Matías shared).

jasmussen commented:
> This is an issue that has come up for a couple of complex blocks and we don't want to block those PRs until this use case is supported. I've gathered what we know so far at #33255 to start thinking about how to solve that issue. While promising, using the elements API requires us to understand first how/whether a UI can be built for them in the post editor, which we don't have designs for yet (this is a concern Matías shared).

Thanks for bringing it up. If we were to ship this PR, would we have any deprecations to handle if we were to refactor it to use the global styles system in the future?

oandregal commented:
> Thanks for bringing it up. If we were to ship this PR, would we have any deprecations to handle if we were to refactor it to use the global styles system in the future?

I don't know what path we'll end up favoring but it's likely that we'd store the data in some consolidated attribute across blocks instead of having custom ones per block. This is the same we did with the existing block supports, and as blocks were migrated to the new system we also added migrations for the attributes. So, I'd say is likely we need to do that in the future, yes.

draganescu commented:
To repeat @georgeh 's comment, let's just make sure this is not available _by default_ in the navigation editor for now. That's all.

Commit:
Fixup whitespace

Commit:
Replacing DOM magic with ref

georgeh commented:
@gwwar can you give this one more look? I'd like to merge this next week but would appreciate any issues you can find.

gwwar commented:
#### Reserving this comment for manual testing

✅ Post-Editor 

https://user-images.githubusercontent.com/1270189/125135442-d4e2b480-e0bd-11eb-81c1-b2d574fd2f36.mp4

✅ Site-Editor

https://user-images.githubusercontent.com/1270189/125135922-aa452b80-e0be-11eb-8123-f1689eb811f0.mp4

✅ Color options not visible in the navigation editor

<img width="1461" alt="Screen Shot 2021-07-09 at 2 06 24 PM" src="https://user-images.githubusercontent.com/1270189/125136083-f7290200-e0be-11eb-830c-0f15092e060f.png">

🤔 I'm seeing some funky navigation styling in the Widgets Editor + 2015 theme, but likely it's not related to color changes here. Probably fine to fix in a follow up.

https://user-images.githubusercontent.com/1270189/125136522-d4e3b400-e0bf-11eb-958e-679d0dec7e14.mp4

gwwar reviewed (approved):
Nice @georgeh ! Great to see this feature added.

I'm giving tentative approval. Ideally we can get E2Es green before merge, though it looks like some E2Es need to be bulletproofed again. 

If folks were happy with the overall color strategy, I think this one is good to go since this is manually testing well.

talldan commented:
Just wanted to mentioned that this PR seems to have introduced an issue with undo/redo that's tracked in https://github.com/WordPress/gutenberg/issues/34564. It's related to the way the `isTopLevelLink` attribute is set using an effect (cc @jasmussen @gwwar).
